### ISSUE TITLE
"the following other types implement trait" needs improvement.

### ISSUE THREAD

**Issue by mejrs** [rust-lang/rust#103822]
tl;dr: this help message is pretty bad, and needs work. See some prior discussion at https://github.com/rust-lang/rust/pull/102752

Let me begin with some examples:
### Showing completely unrelated impls

```rust
pub fn foo() -> Result<u8, u8> {
    let i = 42_i32;
    Ok(i).into()
}
```
```
error[E0277]: the trait bound `Result<u8, u8>: From<Result<i32, _>>` is not satisfied
 --> src/main.rs:4:5
  |
4 |     Ok(i).into()
  |     ^^^^^ ---- required by a bound introduced by this call
  |     |
  |     the trait `From<Result<i32, _>>` is not implemented for `Result<u8, u8>`
  |
  = help: the following other types implement trait `From<T>`:
            <Result<miniz_oxide::MZStatus, miniz_oxide::MZError> as From<&miniz_oxide::StreamResult>>
            <Result<miniz_oxide::MZStatus, miniz_oxide::MZError> as From<miniz_oxide::StreamResult>>
  = note: required for `Result<i32, _>` to implement `Into<Result<u8, u8>>`

For more information about this error, try `rustc --explain E0277`.
```

### Showing hundreds of impls
```rust
42_u8 >> "blah"
```
```
error[E0277]: no implementation for `u8 >> &str`
 --> src/main.rs:2:7
  |
2 | 42_u8 >> "blah"
  |       ^^ no implementation for `u8 >> &str`
  |
  = help: the trait `Shr<&str>` is not implemented for `u8`
  = help: the following other types implement trait `Shr<Rhs>`:
            <&'a i128 as Shr<i128>>
            <&'a i128 as Shr<i16>>
            <&'a i128 as Shr<i32>>
            <&'a i128 as Shr<i64>>
            <&'a i128 as Shr<i8>>
            <&'a i128 as Shr<isize>>
            <&'a i128 as Shr<u128>>
            <&'a i128 as Shr<u16>>
          and 568 others

For more information about this error, try `rustc --explain E0277`.
```
### Cryptic bounds and types
```rust
fn a() -> Result<i32, ()> {
    Err(5)?
}
```
```
error[E0277]: `?` couldn't convert the error to `()`
 --> src/lib.rs:2:11
  |
1 | fn a() -> Result<i32, ()> {
  |           --------------- expected `()` because of this
2 |     Err(5)?
  |           ^ the trait `From<{integer}>` is not implemented for `()`
  |
  = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
  = help: the following other types implement trait `FromResidual<R>`:
            <Result<T, F> as FromResidual<Result<Infallible, E>>>
            <Result<T, F> as FromResidual<Yeet<E>>>
  = note: required for `Result<i32, ()>` to implement `FromResidual<Result<Infallible, {integer}>>`

For more information about this error, try `rustc --explain E0277`.
```

## Discussion

Now, lets look at what these "help" messages tend to have in common:
- they take up a lot of space (especially the second example) and look pretty noisy.
- they require significant effort to parse. To get to "does this look like what I need?" you have to read the message, understand what the mentioned types are and how it might fit in your code. All these things are much harder if you're not familiar with Rust.
- they're more often than not completely useless
- if they're useless, you have to mentally discard all this information you've just loaded.
- because they're usually useless, they introduce fatigue. I don't even look at these anymore, unless I can see at a glance that they're simple like in the case of 
```
   = help: the following other types implement trait `Pattern<'a>`:
             &'b String
             &'b [char; N]
             &'b [char]
             &'b str
             &'c &'b str
             [char; N]
             char
```
but these cases are fairly rare.

## Discussion

Do you agree this is a problem, and what can we do about it?

**Comment by kpreid** [comment]
Idea: if the list of possible impls is very long, as in the `Shr` example, _don't show it at all_ (unless some other heuristic can trim it down by some kind of relevance). The logic here is that if there are many impls, then it is very unlikely a relevant one will be at the top of the truncated list, and the user would be better served by consulting the trait's documentation (which shows all implementations and may have guidance).

The cost of this is that the user would not get the hint "typical impls of this trait look like this", which is occasionally useful, but I think it's a worthwhile tradeoff.

**Comment by tgross35** [comment]
Would it be possible to somehow show what types do implement it, maybe with a flag or something?

When there are 500 types like the example, this info clearly isn't all that useful. But when there are only a handful of possibilities

```
    = help: the following other types implement trait `RxpoTxpo`:
              (NoneT, Pad0, NoneT, NoneT)
              (NoneT, Pad0, Pad2, NoneT)
              (NoneT, Pad0, Pad2, Pad3)
              (Pad0, NoneT, NoneT, NoneT)
              (Pad0, NoneT, Pad2, NoneT)
              (Pad0, NoneT, Pad2, Pad3)
              (Pad1, NoneT, NoneT, NoneT)
              (Pad1, NoneT, Pad2, NoneT)
            and 10 others
```

Come on compiler, show me the options and I'll pick the one I meant!

**Comment by tgross35** [comment]
I know this will never happen, but it would be really awesome to be able to optionally look up the full info that the compiler has about some error messages, especially any that don't fit in the screen now

```
$ cargo c
...
error[E0277]: the trait bound `(Pad0, Pad1, NoneT, NoneT): RxpoTxpo` is not satisfied [id: 36afd9]

$ cargo show-id 36afd9
[full message, context, help, hints, list of available traits - whatever the compiler knows]
```

**Comment by chorman0773** [comment]
I wrote a macro for indexing a tuple using a constant (instead of a literal) that gives an amazing error like this when going out of bounds:
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=99058d44da46eb62b9d08f49b8c44cf6

New WR, btw: 812 other impls. PB attempts to follow.

**Comment by dragomang87** [comment]
> Idea: if the list of possible impls is very long, as in the `Shr` example, _don't show it at all_ (unless some other heuristic can trim it down by some kind of relevance). The logic here is that if there are many impls, then it is very unlikely a relevant one will be at the top of the truncated list, and the user would be better served by consulting the trait's documentation (which shows all implementations and may have guidance).
> 
> The cost of this is that the user would not get the hint "typical impls of this trait look like this", which is occasionally useful, but I think it's a worthwhile tradeoff.

I disagree, what if a user still gets the missing implementation error even though according to the documentation it looks like it should exist? Enabling to see the whole list with some optioin/flag/switch can help understand the problem (e.g. wrong version or some technicality about a generic in question). After having checked the documentation, it would still be nice to be able to confirm that the implementation does not exist or that the implementation that is supposed to be used indeed exists.

**Comment by estebank** [comment]
Current output:

I think this case is appropriate:
```
error[E0277]: the trait bound `Result<u8, u8>: From<Result<i32, _>>` is not satisfied
 --> src/lib.rs:3:11
  |
3 |     Ok(i).into()
  |           ^^^^ the trait `From<Result<i32, _>>` is not implemented for `Result<u8, u8>`
  |
  = note: required for `Result<i32, _>` to implement `Into<Result<u8, u8>>`
```
Few changes, we've gotten better at winnowing down the list of proposed types, to the point where I think it becomes obvious that the problem is that there's no impl for `&str`:
```
error[E0277]: no implementation for `u8 >> &str`
 --> src/main.rs:2:11
  |
2 |     42_u8 >> "blah"
  |           ^^ no implementation for `u8 >> &str`
  |
  = help: the trait `Shr<&str>` is not implemented for `u8`
  = help: the following other types implement trait `Shr<Rhs>`:
            `&u8` implements `Shr<&i128>`
            `&u8` implements `Shr<&i16>`
            `&u8` implements `Shr<&i32>`
            `&u8` implements `Shr<&i64>`
            `&u8` implements `Shr<&i8>`
            `&u8` implements `Shr<&isize>`
            `&u8` implements `Shr<&u128>`
            `&u8` implements `Shr<&u16>`
          and 40 others
```
Work left to be done. I guess that the heuristic for tuples gets triggered in the presence of unit?
```
error[E0277]: `?` couldn't convert the error to `()`
 --> src/lib.rs:2:11
  |
1 | fn a() -> Result<i32, ()> {
  |           --------------- expected `()` because of this
2 |     Err(5)?
  |     ------^ the trait `From<{integer}>` is not implemented for `()`
  |     |
  |     this can't be annotated with `?` because it has type `Result<_, {integer}>`
  |
  = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
  = help: the following other types implement trait `From<T>`:
            `(T, T)` implements `From<[T; 2]>`
            `(T, T, T)` implements `From<[T; 3]>`
            `(T, T, T, T)` implements `From<[T; 4]>`
            `(T, T, T, T, T)` implements `From<[T; 5]>`
            `(T, T, T, T, T, T)` implements `From<[T; 6]>`
            `(T, T, T, T, T, T, T)` implements `From<[T; 7]>`
            `(T, T, T, T, T, T, T, T)` implements `From<[T; 8]>`
            `(T, T, T, T, T, T, T, T, T)` implements `From<[T; 9]>`
          and 4 others
```
I don't think this has changed since reported, but also don't know how to make it better:
```
error[E0277]: get<3>() requires a tuple of length greater than 3
   --> src/lib.rs:103:13
    |
103 |     get::<3,_>(&(0,1,2));
    |             ^ the trait `TupleElement<3>` is not implemented for `({integer}, {integer}, {integer})`
    |
    = help: the following other types implement trait `TupleElement<N>`:
              `(T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19, T20, T21, T22, T23, T24, T25, T26, T27, T28, T29, T30, T31, _T0, _T1, _T2, _T3, _T4, _T5, _T6, _T7)` implements `TupleElement<0>`
              `(T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19, T20, T21, T22, T23, T24, T25, T26, T27, T28, T29, T30, T31, _T0, _T1, _T2, _T3, _T4, _T5, _T6, _T7)` implements `TupleElement<10>`
              `(T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19, T20, T21, T22, T23, T24, T25, T26, T27, T28, T29, T30, T31, _T0, _T1, _T2, _T3, _T4, _T5, _T6, _T7)` implements `TupleElement<11>`
              `(T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19, T20, T21, T22, T23, T24, T25, T26, T27, T28, T29, T30, T31, _T0, _T1, _T2, _T3, _T4, _T5, _T6, _T7)` implements `TupleElement<12>`
              `(T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19, T20, T21, T22, T23, T24, T25, T26, T27, T28, T29, T30, T31, _T0, _T1, _T2, _T3, _T4, _T5, _T6, _T7)` implements `TupleElement<13>`
              `(T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19, T20, T21, T22, T23, T24, T25, T26, T27, T28, T29, T30, T31, _T0, _T1, _T2, _T3, _T4, _T5, _T6, _T7)` implements `TupleElement<14>`
              `(T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19, T20, T21, T22, T23, T24, T25, T26, T27, T28, T29, T30, T31, _T0, _T1, _T2, _T3, _T4, _T5, _T6, _T7)` implements `TupleElement<15>`
              `(T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19, T20, T21, T22, T23, T24, T25, T26, T27, T28, T29, T30, T31, _T0, _T1, _T2, _T3, _T4, _T5, _T6, _T7)` implements `TupleElement<16>`
            and 812 others
note: required by a bound in `get`
   --> src/lib.rs:16:30
    |
16  | pub fn get<const N: usize,T: TupleElement<{N}>>(val: &T) -> &<T as TupleElement<{N}>>::Type{
    |                              ^^^^^^^^^^^^^^^^^ required by this bound in `get`

error[E0277]: get<3>() requires a tuple of length greater than 3
   --> src/lib.rs:103:5
    |
103 |     get::<3,_>(&(0,1,2));
    |     ^^^^^^^^^^^^^^^^^^^^ the trait `TupleElement<3>` is not implemented for `({integer}, {integer}, {integer})`
    |
    = help: the following other types implement trait `TupleElement<N>`:
              `(T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19, T20, T21, T22, T23, T24, T25, T26, T27, T28, T29, T30, T31, _T0, _T1, _T2, _T3, _T4, _T5, _T6, _T7)` implements `TupleElement<0>`
              `(T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19, T20, T21, T22, T23, T24, T25, T26, T27, T28, T29, T30, T31, _T0, _T1, _T2, _T3, _T4, _T5, _T6, _T7)` implements `TupleElement<10>`
              `(T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19, T20, T21, T22, T23, T24, T25, T26, T27, T28, T29, T30, T31, _T0, _T1, _T2, _T3, _T4, _T5, _T6, _T7)` implements `TupleElement<11>`
              `(T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19, T20, T21, T22, T23, T24, T25, T26, T27, T28, T29, T30, T31, _T0, _T1, _T2, _T3, _T4, _T5, _T6, _T7)` implements `TupleElement<12>`
              `(T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19, T20, T21, T22, T23, T24, T25, T26, T27, T28, T29, T30, T31, _T0, _T1, _T2, _T3, _T4, _T5, _T6, _T7)` implements `TupleElement<13>`
              `(T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19, T20, T21, T22, T23, T24, T25, T26, T27, T28, T29, T30, T31, _T0, _T1, _T2, _T3, _T4, _T5, _T6, _T7)` implements `TupleElement<14>`
              `(T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19, T20, T21, T22, T23, T24, T25, T26, T27, T28, T29, T30, T31, _T0, _T1, _T2, _T3, _T4, _T5, _T6, _T7)` implements `TupleElement<15>`
              `(T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19, T20, T21, T22, T23, T24, T25, T26, T27, T28, T29, T30, T31, _T0, _T1, _T2, _T3, _T4, _T5, _T6, _T7)` implements `TupleElement<16>`
            and 812 others
```

**Comment by dragomang87** [comment]
Question: is the list of matching types found or generated? In other words, is the problem that sometimes the matching types are given by intermediate rules that can just explode into a list if there are no restrictions?

The `Pad{0..3}` with `NoneT` example seems to be generated by a rule like "anything tuple of `Pad{0..3}` with `NoneT` with at least one `Pad{0..3}`".

If that is the case, then it might make more sense to first keep all the matching types separated by these intermediate rules and only for those rules that generate 10+ types report the rule with 3 examples as subpoints.

**Comment by estebank** [comment]
@dragomang87 [the logic for mentioning other `impl`s is somewhat involved](https://github.com/rust-lang/rust/blob/fc5af1813307d25a84d633f21e2e53c9376eb547/compiler/rustc_trait_selection/src/error_reporting/traits/fulfillment_errors.rs#L1809), where we different attempts. One is

https://github.com/rust-lang/rust/blob/fc5af1813307d25a84d633f21e2e53c9376eb547/compiler/rustc_trait_selection/src/error_reporting/traits/fulfillment_errors.rs#L1779-L1807

Another is

https://github.com/rust-lang/rust/blob/fc5af1813307d25a84d633f21e2e53c9376eb547/compiler/rustc_trait_selection/src/error_reporting/traits/fulfillment_errors.rs#L1818-L1852

but there's additional winnowing logic to remove candidates after finding them.

We can add more steps to make the heuristics better for our users, absolutely.
